File: scaffold/src/main/java/top/xuqingquan/web/publics/AgentWebJsInterfaceCompat.java
package top.xuqingquan.web.publics;

import android.app.Activity;
import android.os.Build;
import android.webkit.JavascriptInterface;
import androidx.annotation.RequiresApi;
import top.xuqingquan.utils.Timber;
import top.xuqingquan.web.AgentWeb;

import java.lang.ref.WeakReference;

public class AgentWebJsInterfaceCompat {

    private WeakReference<AgentWeb> mReference;
    private WeakReference<Activity> mActivityWeakReference;

    public AgentWebJsInterfaceCompat(AgentWeb agentWeb, Activity activity) {
        mReference = new WeakReference<>(agentWeb);
        mActivityWeakReference = new WeakReference<>(activity);
    }

    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    @JavascriptInterface
    public void uploadFile() {
        uploadFile("*/*");
    }

    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    @JavascriptInterface
    public void uploadFile(String acceptType) {
        Timber.i(acceptType + "  " + mActivityWeakReference.get() + "  " + mReference.get());
        if (mActivityWeakReference.get() != null && mReference.get() != null) {
            AgentWebUtils.showFileChooserCompat(mActivityWeakReference.get(),
                    mReference.get().getWebCreator().getWebView(),
                    null,
                    null,
                    mReference.get().getPermissionInterceptor(),
                    null,
                    acceptType,
                    msg -> {
                        if (mReference.get() != null) {
                            mReference.get().getJsAccessEntrace()
                                    .quickCallJs("uploadFileResult",
                                            msg.obj instanceof String ? (String) msg.obj : null);
                        }
                        return true;
                    }
            );
        }
    }
}

File: scaffold/src/main/java/top/xuqingquan/web/publics/AgentWebJsInterfaceCompat.kt
package top.xuqingquan.web.publics

import android.app.Activity
import android.os.Build
import android.os.Handler
import android.webkit.JavascriptInterface
import android.support.annotation.RequiresApi
import top.xuqingquan.utils.Timber
import top.xuqingquan.web.AgentWeb
import top.xuqingquan.web.nokernel.WebConfig
import java.lang.ref.WeakReference

class AgentWebJsInterfaceCompat(agentWeb: AgentWeb, activity: Activity) {

    private val mReference: WeakReference<AgentWeb> = WeakReference(agentWeb)
    private val mActivityWeakReference: WeakReference<Activity> = WeakReference(activity)

    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    @JavascriptInterface
    @JvmOverloads
    fun uploadFile(acceptType: String = "*/*") {
        Timber.i(acceptType + "  " + mActivityWeakReference.get() + "  " + mReference.get())
        if (mActivityWeakReference.get() != null && mReference.get() != null) {
            if (WebConfig.hasX5()) {
                AgentWebUtils.showFileChooserCompat(mActivityWeakReference.get()!!,
                    mReference.get()!!.x5WebCreator.getWebView()!!, null, null,
                    mReference.get()!!.permissionInterceptor, null,
                    acceptType,
                    Handler.Callback {
                        if (mReference.get() != null) {
                            mReference.get()!!.jsAccessEntrace
                                .quickCallJs(
                                    "uploadFileResult",
                                    if (it.obj is String) {
                                        it.obj as String
                                    } else {
                                        null
                                    }
                                )
                        }
                        true
                    }
                )
            } else {
                AgentWebUtils.showFileChooserCompat(mActivityWeakReference.get()!!,
                    mReference.get()!!.webCreator.getWebView()!!, null, null,
                    mReference.get()!!.permissionInterceptor, null,
                    acceptType,
                    Handler.Callback {
                        if (mReference.get() != null) {
                            mReference.get()!!.jsAccessEntrace
                                .quickCallJs(
                                    "uploadFileResult",
                                    if (it.obj is String) {
                                        it.obj as String
                                    } else {
                                        null
                                    }
                                )
                        }
                        true
                    }
                )
            }
        }
    }
}
